### DISCUSSION TITLE
Add field option for "unique or null"

### DISCUSSION THREAD

**Discussion by thgh** [payloadcms/payload#2072] [Feature Requests & Ideas]
Sometimes a piece of data can be optional, but if it is filled in, it must be unique.

Unique and required
```ts
{
  unique: true,
  required: true
}
```

Unique and not required: null can exist once
```ts
{
  unique: true,
  required: false
}
```

Unique or null: null can exist multiple times
```ts
{
  uniqueOrNull: true,
  required: false
}
```

**Comment by hades200082** [comment]
I'm not sure what use-cases exist for the "null can exist once" option. 

It would seem to me that if unique is true and required is false it should imply "unique or null".

**Reply by thgh** [reply]
Got the same gut feeling, but did actually encounter a usecase for "null can exist once" so I would not make it impossible. Also, it is the default behaviour for unique in postgres.

**Reply by DanRibbens** [reply]
This is valid, I agree!

In Mongo we add the `sparse` option to the index as in the case of localized fields. I noticed the other day that we have no way of allowing `sparse` in the config which is exactly what we need for this requirement.

**Reply by thgh** [reply]
About localized: it seems that if I enable/disable localized, the sparse index stays whatever it was when the field was added. Is this intended behaviour? It caused a bug in a production deployment, did not expect it.

**Reply by DanRibbens** [reply]
Sorry this caused an issue for you @thgh. It has always worked this way. Payload will not remove or change existing indexes from changing the config. As you found out, you have to manually remove them to be re-created automatically when Payload starts again.

**Comment by hades200082** [comment]
Instead of 

```js
{
  uniqueOrNull: true,
  required: false
}
```

Might be better to have `unique` typed as `boolean | { allowNulls: boolean }` and used like this:

```js
{
  unique: {
    allowNulls: true
  },
  required: false
}
```

This retains backwards compatibility with current versions, but adds the new feature without needing a new property.

A later major version _could_ remove the backwards compact and switch it to `false | { allowNulls: boolean }`, the implication bing that if an object is passed then unique is true, and the allowNulls must then be set explicitly.

**Comment by XGZepto** [comment]
It would be sooooo useful to have tbh; is it anywhere on the roadmap yet?

**Comment by ToneseekerMusical** [comment]
This is definitely needed. I've hit a roadblock while working on my website because of this.
